controllers/cluster_controller.go

```diff
@@ -278,70 +301,231 @@ func (r *ClusterReconciler) upgradeControlPlane(ctx context.Context, cluster *ca
+			}
+		}
+		err = r.Client.Update(ctx, kubeadmconfigtemplate)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to update KubeadmConfigTemplate %q", kubeadmconfigtemplate.Name)
+		}
+
+		logger.Info("Changing MachineDeployment k8s version and Machine Template reference to point to new Machine Template", "currentMachineTemplate", machineDeployment.Spec.Template.Spec.InfrastructureRef.Name, "newMachineTemplate", newInfrastructureMachineTemplate.GetName())
+		machineDeployment.Spec.Template.Spec.InfrastructureRef.Name = newInfrastructureMachineTemplate.GetName()
+		machineDeployment.Spec.Template.Spec.Version = to.StringPtr(expectedK8sVersion)
+	}
+
+	err = r.Client.Update(ctx, machineDeployment)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to update MachineDeployment %q", machineDeployment.Name)
+	}
+
+	return nodesWillBeRolled, nil
+}
+
+// upgradeMachinePools will fetch the MachinePools for the given cluster and
+// try to upgrade k8s and OS, setting the right labels.
+func (r *ClusterReconciler) upgradeMachinePools(ctx context.Context, cluster *capi.Cluster, giantswarmRelease *releaseapiextensions.Release) (bool, error) {
+	clusterMachinePools, err := r.getSortedClusterMachinePools(ctx, cluster)
+	if err != nil {
+		return false, err
+	}
+
+	for _, machinePool := range clusterMachinePools {
+		logger := r.Log.WithValues("cluster", cluster.Name, "machinepool", machinePool.Name)
+
+		// If a MachinePool is not ready we exit because we don't want to
 		// trigger an upgrade while something is wrong or another upgrade is in
 		// progress.
 		if !isReady(&machinePool) {
-			logger.Info("The MachinePool is not ready. Let's wait before trying to continue upgrading workers")
-			return nil
-		} else if currentMachineImage != expectedMachineImage || *machinePool.Spec.Template.Spec.Version != expectedK8sVersion || machinePool.Labels[CAPIWatchFilterLabel] != expectedCAPIVersion {
-			logger.Info("The MachinePool is out of date. Updating the MachinePool and its infrastructure Machine Pool", "expectedK8sVersion", expectedK8sVersion, "currentK8sVersion", *machinePool.Spec.Template.Spec.Version, "expectedMachineImage", expectedMachineImage, "currentMachineImage", currentMachineImage, "expectedWatchFilterLabel", expectedCAPIVersion, "currentWatchFilterLabel", machinePool.Labels[CAPIWatchFilterLabel])
-			machinePool.Labels[CAPIWatchFilterLabel] = expectedCAPIVersion
-			machinePool.Spec.Template.Spec.Version = &expectedK8sVersion
-			err = r.Client.Update(ctx, &machinePool)
-			if err != nil {
-				return err
-			}
+			logger.Info("The MachinePool is not ready. Let's wait before trying to continue upgrading machine pools")
+			return true, nil
+		}
+
+		nodesWillBeRolled, err := r.upgradeMachinePool(ctx, cluster, &machinePool, giantswarmRelease)
+		if err != nil {
+			return false, err
+		}
 
-			//ready, err := external.IsReady(infraMachinePool)
-			//if !ready {
-			//	return ctrl.Result{}, nil
-			//}
+		if nodesWillBeRolled {
```

Comment:
Could be just: `return nodesWillBeRolled, nil`

Reply:
No because we don't want to return/exit if there are no nodes being rolled.